File: 2501~3000/2642. Design Graph With Shortest Path Calculator/main.py
"""
Source: https://leetcode.com/problems/design-graph-with-shortest-path-calculator/
Date: 2023/4/22
Skill: Floyd–Warshall algorithm
Ref:
Runtime: 7271 ms, faster than 6.20%
Memory Usage: 17.1 MB, less than 76.27%
Time complexity:
Space complexity:
Constraints:
    1 <= n <= 100
"""

import math
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Graph:

    def __init__(self, n: int, edges: List[List[int]]):
        self.dp, self.n = [[math.inf for _ in range(n)] for _ in range(n)], n
        for i in range(n):
            self.dp[i][i] = 0
        for edge in edges:
            self.dp[edge[0]][edge[1]] = edge[2]

        for i in range(n):
            for j in range(n):
                for k in range(n):
                    self.dp[j][k] = min(self.dp[j][k], self.dp[j][i] + self.dp[i][k])

    def addEdge(self, edge: List[int]) -> None:
        if edge[2] < self.dp[edge[0]][edge[1]]:
            self.dp[edge[0]][edge[1]] = edge[2]
            for i in range(self.n):
                for j in range(self.n):
                    self.dp[i][j] = min(self.dp[i][j], self.dp[i][edge[0]] + edge[2] + self.dp[edge[1]][j])

    def shortestPath(self, node1: int, node2: int) -> int:
        return self.dp[node1][node2] if self.dp[node1][node2] < math.inf else -1


# Your Graph object will be instantiated and called as such:
# obj = Graph(n, edges)
# obj.addEdge(edge)
# param_2 = obj.shortestPath(node1,node2)


if __name__ == "__main__":
    s = Graph()

File: 2501~3000/2642. Design Graph With Shortest Path Calculator/main.cpp
/**
 * Source: t.ly/WNyEM
 * Date: 2023/11/11
 * Skill:
 * Runtime: 208 ms, faster than 72.27% of C++ online submissions
 * Memory Usage: 112.12 MB, less than 21.57% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */


#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;

#define ll long long

typedef pair<int, int> pairs;

const int MX = 101;


class Graph {
    int visited[MX] = {0};
    vector<pairs> next[MX];
public:
    Graph(int n, vector<vector<int>>& edges) {
        for (auto &e: edges) {
            next[e[0]].push_back({e[1], e[2]});
        }
    }

    void addEdge(vector<int> edge) {
        next[edge[0]].push_back({edge[1], edge[2]});
    }

    int shortestPath(int node1, int node2) {
        memset(visited, 0, sizeof(visited));
        priority_queue<pairs, vector<pairs>, greater<pairs>> pq;
        pq.push({0, node1});
        while (!pq.empty()) {
            auto [dis, pos] = pq.top();
            pq.pop();
            if (pos == node2) return dis;
            if (visited[pos] == 1) continue;
            visited[pos] = 1;
            for (auto &i: next[pos]) {
                if (visited[i.first] == 1) continue;
                pq.push({i.second + dis, i.first});
            }
        }
        return -1;
    }
};

/**
 * Your Graph object will be instantiated and called as such:
 * Graph* obj = new Graph(n, edges);
 * obj->addEdge(edge);
 * int param_2 = obj->shortestPath(node1,node2);
 */


int main() {

}